google/cloud/pubsub/internal/publisher_tracing_connection_test.cc

```diff
@@ -15,35 +15,132 @@
+#include "gtest/gtest.h"
 #include <gmock/gmock.h>
+#include <opentelemetry/trace/semantic_conventions.h>
 
 namespace google {
 namespace cloud {
 namespace pubsub_internal {
 namespace {
 
+auto constexpr kErrorCode = static_cast<int>(StatusCode::kAborted);
+
 using ::google::cloud::pubsub::PublisherConnection;
+using ::google::cloud::pubsub::Topic;
 using ::google::cloud::pubsub_internal::MakePublisherTracingConnection;
 using ::google::cloud::pubsub_mocks::MockPublisherConnection;
 using ::google::cloud::testing_util::InstallSpanCatcher;
 using ::google::cloud::testing_util::OTelAttribute;
 using ::google::cloud::testing_util::SpanHasAttributes;
 using ::google::cloud::testing_util::SpanHasInstrumentationScope;
 using ::google::cloud::testing_util::SpanKindIsClient;
+using ::google::cloud::testing_util::SpanKindIsProducer;
 using ::google::cloud::testing_util::SpanNamed;
 using ::google::cloud::testing_util::SpanWithStatus;
+using ::google::cloud::testing_util::StatusIs;
 using ::google::cloud::testing_util::ThereIsAnActiveSpan;
+using ::testing::AllOf;
+using ::testing::ElementsAre;
+
+TEST(PublisherTracingConnectionTest, PublishSpanOnSuccess) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_shared<MockPublisherConnection>();
+  EXPECT_CALL(*mock, Publish)
+      .WillOnce([&](pubsub::PublisherConnection::PublishParams const&) {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return make_ready_future(StatusOr<std::string>("test-id-0"));
+      });
+  auto connection = MakePublisherTracingConnection(
+      Topic("test-project", "test-topic"), std::move(mock));
+
+  auto response = connection
+                      ->Publish({pubsub::MessageBuilder{}
+                                     .SetData("test-data-0")
+                                     .SetOrderingKey("ordering-key-0")
+                                     .Build()})
+                      .get();
+
+  EXPECT_THAT(response, StatusIs(StatusCode::kOk));
+  EXPECT_THAT(
+      span_catcher->GetSpans(),
+      ElementsAre(AllOf(
+          SpanHasInstrumentationScope(), SpanKindIsProducer(),
+          SpanNamed("projects/test-project/topics/test-topic send"),
+          SpanWithStatus(opentelemetry::trace::StatusCode::kOk),
+          SpanHasAttributes(
+              OTelAttribute<std::string>(sc::kMessagingSystem, "pubsub"),
+              OTelAttribute<std::string>(
+                  sc::kMessagingDestinationName,
+                  "projects/test-project/topics/test-topic"),
+              OTelAttribute<std::string>(sc::kMessagingDestinationTemplate,
+                                         "topic"),
+              OTelAttribute<std::string>("messaging.pubsub.ordering_key",
+                                         "ordering-key-0"),
+              OTelAttribute<int>("gcloud.status_code", 0),
+              OTelAttribute<size_t>("messaging.message.total_size_bytes", 45),
```

Comment:
nit: I don't think `std::size_t` is a consistently sized attribute type: https://github.com/open-telemetry/opentelemetry-cpp/blob/cdbc90abf3b7170f898439fd34126469c42686c6/sdk/include/opentelemetry/sdk/common/attribute_utils.h#L30-L44

Maybe use `std::uint64_t`?

Reply:
It isn't. I tried `std::uint64_t` and it failed the m32 build. 

I'll try`std::uint32_t`.

Reply:
From one of the failing builds
```
Step #3:   Actual: { [messaging.message_id=std::string:test-id-0, messaging.message.total_size_bytes=std::uint64_t:45 ]} 

Step #3: matcher #5: has a first field that is equal to "messaging.message.total_size_bytes", and has a second field that is a variant<> with value of type 'unsigned int' and the value is equal to 45
```
Gonna go back to `size_t`.

Let me know if there's a better way to do this. Maybe we could check the architecture size then set the type at run time?